packages/editor/src/editor.ts

```diff
@@ -0,0 +1,236 @@
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import { once } from 'events';
+import { Readable } from 'stream';
+import spawn from 'cross-spawn';
+import { v4 as uuidv4 } from 'uuid';
+import { signatures, ShellPlugin, ShellInstanceState, TypeSignature } from '@mongosh/shell-api';
+
+import type { MongoshBus } from '@mongosh/types';
+
+const beautify = require('js-beautify').js;
+
+interface EditorSession {
+  cmd: string;
+  args: string[];
+  content: string;
+  code: string;
+  tmpDoc: string;
+}
+
+export interface EditorOptions {
+  input: Readable;
+  vscodeDir: string;
+  tmpDir: string;
+  instanceState: ShellInstanceState;
+  makeMultilineJSIntoSingleLine: any;
+  loadExternalCode: any;
+}
+
+export class Editor implements ShellPlugin {
+  _input: Readable;
+  _vscodeDir: string;
+  _tmpDir: string;
+  _instanceState: ShellInstanceState;
+  _makeMultilineJSIntoSingleLine: any;
+  _loadExternalCode: any;
+  require: any;
+  _editorSession: EditorSession;
+  print: (...args: any[]) => Promise<void>;
+
+  constructor({ input, vscodeDir, tmpDir, instanceState, makeMultilineJSIntoSingleLine, loadExternalCode }: EditorOptions) {
+    this._input = input;
+    this._vscodeDir = vscodeDir;
+    this._tmpDir = tmpDir;
+    this._instanceState = instanceState;
+    this._makeMultilineJSIntoSingleLine = makeMultilineJSIntoSingleLine;
+    this._loadExternalCode = loadExternalCode;
+    this._editorSession = { cmd: '', args: [], content: '', code: '', tmpDoc: '' };
```

Comment:
Is there a reason to store the state here, instead of locally in the functions where it is used? Since the state (other than maybe the content from the of the last `edit` command) does not span across multiple `runEditCommand()` calls, that would feel a bit cleaner, I think